aeron-client/src/main/c/concurrent/aeron_executor.c

```diff
@@ -0,0 +1,310 @@
+};
+
+#define USE_RETURN_QUEUE(_e) (NULL == (_e)->on_execution_complete)
+
+aeron_executor_task_t *aeron_executor_task_acquire(
+    aeron_executor_t *executor,
+    aeron_executor_task_on_execute_func_t on_execute,
+    aeron_executor_task_on_complete_func_t on_complete,
+    void *clientd,
+    bool shutdown)
+{
+    aeron_executor_task_t *task;
+
+    if (aeron_alloc((void **)&task, sizeof(aeron_executor_task_t)) < 0)
+    {
+        AERON_APPEND_ERR("%s", "");
+        return NULL;
+    }
+
+    task->executor = executor;
+    task->on_execute = on_execute;
+    task->on_complete = on_complete;
+    task->clientd = clientd;
+    task->result = -1;
+    task->queue_node = NULL;
+    task->shutdown = shutdown;
+
+    return task;
+}
+
+void aeron_executor_task_release(aeron_executor_task_t *task)
+{
+    if (NULL != task)
+    {
+        aeron_linked_queue_node_delete(task->queue_node);
+        aeron_free(task);
+    }
+}
+
+static void *aeron_executor_dispatch(void *arg)
+{
+    aeron_executor_t *executor = (aeron_executor_t *)arg;
+
+    aeron_thread_set_name("aeron_executor");
+
+    aeron_executor_task_t *task;
+    aeron_linked_queue_node_t *node;
+    bool shutdown = false;
+
+    do
+    {
+        task = (aeron_executor_task_t *)aeron_blocking_linked_queue_take_ex(&executor->queue, &node);
+
+        if (NULL == task)
+        {
+            continue;
+        }
+
+        shutdown = task->shutdown;
+
+        task->queue_node = node;
+        task->result = (NULL == task->on_execute) ? 0 : task->on_execute(task->clientd, executor->clientd);
+
+        if (task->result < 0)
+        {
+            task->errcode = aeron_errcode();
+            memcpy(task->errmsg, aeron_errmsg(), strlen(aeron_errmsg()));
+            aeron_err_clear();
+        }
+
+        if (USE_RETURN_QUEUE(executor))
+        {
+            aeron_blocking_linked_queue_offer_ex(&executor->return_queue, task, node);
+        }
+        else if (shutdown)
+        {
+            aeron_executor_task_release(task);
+        }
+        else
+        {
+            executor->on_execution_complete(task, executor->clientd);
+        }
+    }
+    while (false == shutdown);
+
+    return NULL;
+}
+
+int aeron_executor_init(
+    aeron_executor_t *executor,
+    bool async,
+    aeron_executor_on_execution_complete_func_t on_execution_complete,
+    void *clientd)
+{
+    executor->async = async,
+    executor->on_execution_complete = on_execution_complete;
+    executor->clientd = clientd;
+
+    if (async)
+    {
+        if (USE_RETURN_QUEUE(executor))
+        {
+            if (aeron_blocking_linked_queue_init(&executor->return_queue) < 0)
+            {
+                AERON_APPEND_ERR("%s", "");
+                return -1;
+            }
+        }
+
+        if (aeron_blocking_linked_queue_init(&executor->queue) < 0)
+        {
+            AERON_APPEND_ERR("%s", "");
+            return -1;
+        }
+
+        pthread_attr_t attr;
```

Comment:
Should be `aeron_thread_attr`

Reply:
I don't see an aeron_thread_attr defined anywhere.  Should I add a typedef in aeron_thread.h?  I think I was just copying what I saw in aeron_agent.c in aeron_agent_start().